Supporting changes for Cloud email invites

timothyb89 opened the pull request:
This adds a number of changes to support email user invites for Cloud.

![image](https://github.com/gravitational/teleport.e/assets/3459954/e70ecc6e-b7ae-49a7-847b-66d13738e226)

* Adds a new public `/preset-roles` endpoint that just returns the preset roles for the current version of Teleport. It's unauthenticated and intended for use in the new onboarding invitation UI. I don't think this has security implications  beyond what we already return in the `ping` replies, since those include the Teleport version.
* Includes role descriptions in the JSON responses sent to the UI
* Makes various tweaks to our `react-select` wrapper, and adds a Creatable variant for email entry.
* Adds a new `requiresEmailLike` validation rule, and adds more typing to the rules
* Adds support for providing an invite collaborators element to replace the usual "Create User" UI from teleport.e
* Adds support for showing the invite collaborators element during onboard, if a flag is set as a query parameter in the invite URL.

Note that gravitational/teleport.e#2286 is needed to test the changes. This PR must be merged before the `teleport.e` PR, then another PR here should bump the .e ref.

See also:
 * https://github.com/gravitational/cloud/issues/5137
 * .e changes: https://github.com/gravitational/teleport.e/pull/2286
 * Sales Center RPC: https://github.com/gravitational/cloud/pull/5970

timothyb89 committed:
Add WIP implementation of Teleport email invites

This adds a WIP impl of Teleport email invites. Requires a compatible
Enterprise build and Cloud API.

timothyb89 committed:
Bump e ref and add new validation rule

timothyb89 committed:
Various improvements to enable Cloud email invites

* Add description to UI role resources
* Expose various new react-select options
* Add new FieldSelectCreatable
* Add some typing for validation rules
* Tweak invite button for Cloud to use email UI instead of showing
  both buttons
* Partial implementation for onboarding invites

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/teleport-invites

timothyb89 committed:
Add support for Cloud collaborator invites during onboarding

This adds various changes to enable showing the invite collaborators
form during initial user onboarding.

 * Adds a `?initial` URL query parameter for the UI to signify the
   first user; Cloud will append this to invite appropriate invite
   links.
 * Added a new ratelimited public endpoint to return a list of preset
   roles. This just exposes static data available otherwise available
   in Git and that could be obtained from the public Teleport version
   shown in ping responses already.

timothyb89 committed:
Update e ref for the invite-collaborators branch

timothyb89 committed:
Honor the `inputId` parameter if set

timothyb89 committed:
bump e ref

timothyb89 committed:
Improve typing for `requiredEmailLike` to add a error category

The `kind` field can allow the UI to group errors together if several
invalid emails are entered.

timothyb89 committed:
bump e ref

kimlisa commented on web/packages/teleport/src/Welcome/NewCredentials/NewCredentials.tsx:
```diff
@@ -66,6 +66,9 @@ export function NewCredentials(props: NewCredentialsProps) {
     setDisplayOnboardingQuestionnaire = false,
     Questionnaire = undefined,
+    displayInviteCollaborators = false,
+    setDisplayInviteCollaborators = false,
```
i'd avoid naming things `setXXX` unless it's a react hook state setter (that's how it's normally named `setXXX`), not sure how to give rename suggestions b/c i don't really understand it yet

timothyb89 replied:
Right, sorry, I was following the pattern from `setDisplayOnboardingQuestionnaire` but I think I can just as well set it to `null`.

kimlisa commented on web/packages/teleport/src/Welcome/NewCredentials/types.ts:
```diff
@@ -60,6 +66,13 @@ export type NewCredentialsProps = UseTokenState & {
+
+  // support for E's invite collaborators at onboarding
+  displayInviteCollaborators?: boolean;
+  setDisplayInviteCollaborators?: (bool: boolean) => void;
```
oh so it is a setter, i think setting a default to false was a mistake then? (see previous comment)

timothyb89 replied:
Yup, I'll set it to null by default instead (per the type signature I suppose).

kimlisa reviewed (commented):
first pass, will continue after a few questions:

- not sure if i'm doing something wrong (or misunderstanding something), but this these changes should only affect cloud users right? (i'm running enterprise and i'm seeing these changes, i can't add a user through the web ui)

- what if a user wants to add a non-email user?

- anyway you can spin a cluster with all your changes so i can test it? 🙏 


also there is a weird flash when rendering the dialog because of this height:

<img width="639" alt="image" src="https://github.com/gravitational/teleport/assets/43280172/2a0d796d-051a-40be-aeb9-7dab8a38a181">

timothyb89 commented:
> first pass, will continue after a few questions:
> 
> * not sure if i'm doing something wrong (or misunderstanding something), but this these changes should only affect cloud users right? (i'm running enterprise and i'm seeing these changes, i can't add a user through the web ui)

Yes, it should only affect Cloud. I'll look into why you're seeing it on Enterprise.

> * what if a user wants to add a non-email user?

The decision was that non-email users can no longer be created on Cloud. If someone wants to add one, they'll need to use `tctl`. I think the plan is to see what sort of customer feedback we get on that change.

>  * anyway you can spin a cluster with all your changes so i can test it? 🙏

Hmm, that's a good idea, I think it'll be possible once the changes in Sales Center are deployed (hopefully soon): https://github.com/gravitational/cloud/pull/5970.

timothyb89 commented:
Various TODOs needed before this can be merged:

- [ ] Add unit tests
- [ ] Remove `e` ref bump - the changes are interdependent so the supporting changes will need to be merged sans the `e` bump, then we'll bump it via a separate PR.

mcbattirola commented on web/packages/teleport/src/Users/Users.tsx:
```diff
@@ -49,16 +49,31 @@ export function Users(props: State) {
     onReset,
+    onStartInviteCollaborators,
+    onInviteCollaboratorsClose,
+    inviteCollaboratorsOpen,
```
We can destructure `inviteCollaborators` here to follow the other props.

timothyb89 replied:
Sure, I'd done it this way since it made the type checker angry, turns out intrinsic elements just need to be capitalized? :facepalm:  Fixed now at least.

mcbattirola commented on web/packages/teleport/src/Users/Users.tsx:
```diff
@@ -49,16 +49,31 @@ export function Users(props: State) {
+                Create New User
+              </ButtonPrimary>
+            )}
+            {props.inviteCollaborators && (
```
If I'm reading correctly, this will display `Enroll Users` for any enterprise customers since we are always sending this on enterprise features (https://github.com/gravitational/teleport.e/blob/f32252614746014c5a9136cd34a9bf3ee2dc5d8f/web/teleport/src/features.tsx#L376)

You could use the `isCloud` flag on `web/packages/teleport/src/config.ts` to show/hide this instead.

timothyb89 replied:
Yup, that did the trick, thanks!

timothyb89 committed:
Destructure the InviteCollaborators component sanely

timothyb89 committed:
Set `setDisplayInviteCollaborators` to `null` instead of `false`

timothyb89 committed:
Split `FieldSelectCreatable` into its own file

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/teleport-invites

timothyb89 commented on web/packages/shared/components/FieldSelect/FieldSelect.tsx:
```diff
@@ -75,6 +84,72 @@ export default function FieldSelect({
   );
 }
 
+export function FieldSelectCreatable({
```
Sure, I'll sync this with Cloud once it's merged.

timothyb89 commented:
A few non-threaded changes:
 * The flash of content on the dialog is fixed (the loading spinner no longer has a minimum height)
 * The Enroll Users button (and new feature) is no longer shown for Enterprise
 * The `e` ref update is removed, so for testing you'll need to check out https://github.com/gravitational/teleport.e/pull/2286

timothyb89 committed:
Fix lint

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/teleport-invites

timothyb89 committed:
add story for SelectCreatable

timothyb89 committed:
Add tests for `requiredEmailLike`

timothyb89 committed:
Rename `initial` flag to `invite`

Renaming the flag will hopefully clarify the intent.

timothyb89 committed:
Add tests for invite collaborators feedback and users rendering

timothyb89 committed:
Add rendering test for the invite collaborators card

timothyb89 committed:
Clean up lints

timothyb89 commented:
More notes: I think we've now got reasonable testing / storybook coverage for this PR (though I have some more work to do in the corresponding .e change).

Also, with the [cloud changes merged](https://github.com/gravitational/cloud/pull/5970), I've now got a test cluster running this change, so DM me if you'd like an account.

Additionally, per some feedback I've renamed the `?initial` parameter to `?invite` to make it what it does a little more explicit.

kimlisa reviewed (approved):
thanks a lot of the changes and the cluster 👍 

most of my comments are pretty minor so i'll approve this.

I think some of the feedback for below is actually being affected in enteprise:

some rule validation is kinda broken, it sometimes allows dups, sometimes it doesn't, and when it doesn't it duplicates error message

this is an error after making an api call (allows dups):

<img width="602" alt="image" src="https://github.com/gravitational/teleport/assets/43280172/5b1c7d4d-a962-4678-86c6-5ecd20fdce2a">

this prevents me, but the error msg's are duplciated:

<img width="530" alt="image" src="https://github.com/gravitational/teleport/assets/43280172/2af32a24-6048-42b0-8c34-37c370935009">

nit:

<img width="735" alt="image" src="https://github.com/gravitational/teleport/assets/43280172/4398076c-2bff-4dce-bd43-2df9e5024224">

probably caused by some kind of initial loading, i suggest trying out this css property: [visibility](https://developer.mozilla.org/en-US/docs/Web/CSS/visibility). you'd probably have to lay the loader on top of this block

mcbattirola reviewed (approved):
LGTM once Lisa's comments are addressed

timothyb89 commented on web/packages/shared/components/Select/types.ts:
```diff
@@ -45,6 +49,7 @@ export type Props = {
   onInputChange?(value: string, actionMeta: ActionMeta): void;
   // Whether or not the element is on an elevated platform (such as a dialog).
   elevated?: boolean;
+  stylesConfig?: StylesConfig;
```
It does do that, but it also lets us highlight items selectively. That's how specific invalid addresses get highlighted in red.

timothyb89 commented on web/packages/shared/components/FieldSelect/types.tsx:
Given `defaultRule` is also not a type, I've just renamed this file to `shared.tsx`. It still has `LabelTip` which needs .tsx.

timothyb89 commented on web/packages/shared/components/FieldSelect/types.tsx:
```diff
@@ -0,0 +1,23 @@
+
+export const defaultRule = () => () => ({ valid: true });
+
+export const LabelTip = ({ text }) => (
+  <span
+    css={{ fontWeight: 'normal', textTransform: 'none' }}
+  >{` - ${text}`}</span>
+);
```
(duplicating the other comment) I'll rename the file outright since the other function is also not a type.

timothyb89 committed:
Rename types.tsx -> shared.tsx

timothyb89 committed:
Relocate invite constant to `Welcome/const.ts`

timothyb89 committed:
Split `SelectCreatable` into its own story

timothyb89 committed:
Clarify SelectCreatable story

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/teleport-invites

timothyb89 committed:
Simplify story; fix lint

timothyb89 commented on web/packages/shared/components/Select/Select.story.tsx:
```diff
@@ -112,3 +113,45 @@ function SelectDark({ value, onChange, options }) {
   );
 }
+
+function SelectCreatableDefault() {
```
I've split it into its own dedicated story and added labels. It definitely works on my machine, though in this mode (without a custom `keyDown` handler like we use in `InviteCollaboratorsForm`) input is a bit goofy, just due to react-select's defaults. I added some comments to explain what you might want to do to actually use the component in a  real world setting.

timothyb89 commented:
> some rule validation is kinda broken, it sometimes allows dups, sometimes it doesn't, and when it doesn't it duplicates error message

I've added a new rule to ensure there's no duplicates within the entered field (we only checked against cluster users), and refactored the validation to aggregate all the various error messages so the strings aren't too verbose. I also noticed we didn't error highlight for user exists/duplicates, so added a new validator for that as well. 

It now looks like this (hopefully worst case):
![image](https://github.com/gravitational/teleport/assets/3459954/fb68cb79-dd60-47c8-a81e-3c6c81ccb30f)

(these changes are mostly in the .e PR)

> probably caused by some kind of initial loading, i suggest trying out this css property: [visibility](https://developer.mozilla.org/en-US/docs/Web/CSS/visibility). you'd probably have to lay the loader on top of this block

I've now set it to use `visibility: hidden` while loading (also in the .e PR).

timothyb89 committed:
Fix type checker failure

zmb3 commented on lib/web/apiserver.go:
```diff
@@ -767,6 +767,7 @@ func (h *Handler) bindDefaultEndpoints() {
 	h.POST("/webapi/roles", h.WithAuth(h.upsertRoleHandle))
 	h.PUT("/webapi/roles/:name", h.WithAuth(h.upsertRoleHandle))
 	h.DELETE("/webapi/roles/:name", h.WithAuth(h.deleteRole))
+	h.GET("/webapi/preset-roles", h.WithUnauthenticatedHighLimiter(h.getPresetRoles))
```
I don't think we typically put - characters in URL paths.

timothyb89 replied:
I'll rename it to `presetroles`

zmb3 commented on lib/web/resources.go:
```diff
@@ -150,6 +151,15 @@ func upsertRole(ctx context.Context, clt resourcesAPIGetter, content, httpMethod
+// via the public ping endpoint.
+func (h *Handler) getPresetRoles(w http.ResponseWriter, r *http.Request, p httprouter.Params) (interface{}, error) {
+	presets := auth.GetPresetRoles()
+	return ui.NewRoles(presets)
```
Are we returning the full content of the role to unauthenticated users?

timothyb89 replied:
Yes, since this is used during onboarding (before credentials are created) as an alternative to fetching the actual roles on the cluster, which would reveal specific information and requires auth. For what it's worth, `auth.GetPresetRoles()` does not query the database - it is the same code that generates the roles that _would_ be inserted, and will always return the same static content for a given binary, regardless of any changes users make to those roles.

These are hard-coded and correspond directly to the Teleport version, which is already available to unauthenticated users via the `ping` endpoint, so an attacker could get the same information by cloning our git repo and checking out the appropriate tag. Hopefully if/when we decide to mask the Teleport version (and this new endpoint) we'll have revised the onboarding flow to be able to query for actual cluster roles.

zmb3 replied:
Thanks for the detail. Sounds fine to me.

timothyb89 committed:
Rename `preset-roles` endpoint to `presetroles` to follow API conventions
